src/PointwiseFunctions/GeneralRelativity/GeneralizedHarmonic/Ricci.cpp

```diff
@@ -0,0 +1,167 @@
+
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tags/TempTensor.hpp"
+#include "DataStructures/TempBuffer.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Utilities/ContainerHelpers.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+// IWYU pragma: no_forward_declare Tensor
+
+/// \cond
+namespace GeneralizedHarmonic {
+template <size_t VolumeDim, typename Frame, typename DataType>
+void spatial_ricci_tensor(
+    const gsl::not_null<tnsr::ii<DataType, VolumeDim, Frame>*> ricci,
+    const tnsr::iaa<DataType, VolumeDim, Frame>& phi,
+    const tnsr::ijaa<DataType, VolumeDim, Frame>& deriv_phi,
+    const tnsr::II<DataType, VolumeDim, Frame>&
+        inverse_spatial_metric) noexcept {
+  destructive_resize_components(ricci, get_size(get<0, 0, 0>(phi)));
+
+  TempBuffer<tmpl::list<::Tags::TempIjj<0, VolumeDim, Frame, DataType>,
+                        ::Tags::TempijK<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<0, VolumeDim, Frame, DataType>,
+                        ::Tags::Tempi<1, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<0, VolumeDim, Frame, DataType>,
+                        ::Tags::TempI<1, VolumeDim, Frame, DataType>>>
+      local_buffer(get_size(get<0, 0>(inverse_spatial_metric)));
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_Ijj =
+      get<::Tags::TempIjj<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = i; j < VolumeDim; ++j) {  // Symmetry
+        spatial_phi_Ijj.get(k, i, j) =
+            0.5 * inverse_spatial_metric.get(k, 0) * phi.get(0, 1 + i, 1 + j);
+        for (size_t l = 1; l < VolumeDim; ++l) {
+          spatial_phi_Ijj.get(k, i, j) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(l, 1 + i, 1 + j);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_ijK =
+      get<::Tags::TempijK<0, VolumeDim, Frame, DataType>>(local_buffer);
+  for (size_t k = 0; k < VolumeDim; ++k) {
+    for (size_t i = 0; i < VolumeDim; ++i) {
+      for (size_t j = 0; j < VolumeDim; ++j) {
+        spatial_phi_ijK.get(i, j, k) =
+            0.5 * inverse_spatial_metric.get(k, 0) * phi.get(i, 1 + j, 1);
+        for (size_t l = 1; l < VolumeDim; ++l) {
+          spatial_phi_ijK.get(i, j, k) +=
+              0.5 * inverse_spatial_metric.get(k, l) * phi.get(i, 1 + j, 1 + l);
+        }
+      }
+    }
+  }
+
+  // New variable to avoid recomputing in nested loops
+  auto& spatial_phi_trace_second_third_indices =
+      get<::Tags::Tempi<0, VolumeDim, Frame, DataType>>(local_buffer);
+  auto& spatial_phi_trace_first_second_indices =
+      get<::Tags::Tempi<1, VolumeDim, Frame, DataType>>(local_buffer);
```

Comment:
Should these have `half` in their names?

Reply:
yes, renamed